### DISCUSSION TITLE
Theia license not really attractive for contributors

### DISCUSSION THREAD

**Discussion by nimo23** [eclipse-theia/theia#14502] [Q&A] [answered]
It is a pity that this project has been switched from **Apache License** to **EPL**.  

I will share with you my opinion why I see it this way:

- it's to the detriment of the project because it's preventing so many potential contributors from contributing. The proof is in the air: look at the number of contributors in `VS-code `and look at the number of contributors in `Theia`. Btw, any potential contributor needs to sign a https://github.com/eclipse-theia/theia/blob/0556d2a4ebcf0024e96a1e54ae3450f5e6ea24d6/CONTRIBUTING.md#eclipse-contributor-agreement (because of https://www.eclipse.org/legal/epl-2.0/faq/#h.973c2qv6zg27). Contributor who just wants to do a quick PR thing, the fun is spoiled because of this kind of bureaucracy. **MIT** or **Apache** license is better and would appeal more contributors. 

-  Correct me if I am wrong: But I currently see `Theia` itself as a (modified) fork of `VS Code` (with added extensions) - Imagine if `VS Code` was also licensed with `EPL`, would `Theia` even exist? Imagine that if, for example, someone **just wants to change some CSS or React code from the Theia source code** and add his own proprietary things, he is forced to release it explicitly and in an accessible way as open source  (see https://www.eclipse.org/legal/epl-2.0/faq/#h.2c51ohn0wg82) - this makes easy distribution of his proprietary software much more difficult (or nearly impossible). 

Don't you think that returning to the **Apache License** or **MIT** would be better for this project? If not, why? What is the benefit of **EPL** for the project and its contributors? I may not fully understand, as I'm not a lawyer, but just someone who is considering investing my time in either `Theia` or `VS Code` (including my possible contributions).

**Comment by msujew** [comment]
Hey @nimo23,

we always appreciate input on the project. There are quite a few misunderstandings that I want to quickly clear up before getting to the license question:

> The proof is in the air: look at the number of contributors in VS-code and look at the number of contributors in Theia.

No offense, but that's comparing oranges to apples - our target audience for Theia is much smaller than VS Code's. We're honestly really happy about the amount of contributors that are actively contributing to the project.

> Contributor who just wants to do a quick PR thing, the fun is spoiled because of this kind of bureaucracy. MIT or Apache license is better and would appeal more contributors.

Well, but the same applies to VS Code. All Microsoft GitHub projects enforce a contributor license agreement (CLA) - see literally any PR on the VS Code repo - that ensure that Microsoft has rights to the contributed code. This is a requirement to essentially every large open source project, as it ensures legal rights to all code in the repository for the owning entity (i.e. Eclipse Foundation in our case or Microsoft for VS Code). This is done as a legal liability measure to ensure that adopters of the respective frameworks don't accidentally violate any intellectual property rights by using the code. Note that CLAs are relatively unrelated to the chosen license. We have to use the same CLA for MIT licensed projects.

> But I currently see Theia itself as a (modified) fork of VS Code (with added extensions) - Imagine if VS Code was also licensed with EPL, would Theia even exist?

Theia was never a fork of VS Code and is not designed as such. We probably share less than 2% of code (mostly to keep compatibility regarding APIs).

> Imagine that if, for example, someone just wants to change some CSS or React code from the Theia source code and add his own proprietary things, he is forced to release it explicitly and in an accessible way as open source (see https://www.eclipse.org/legal/epl-2.0/faq/#h.2c51ohn0wg82) - this makes easy distribution of his proprietary software much more difficult (or nearly impossible).

Theia (as a framework) in it's core is not designed to be forked. Instead, it features a custom extension mechanism (not to be confused with the VS Code extension mechanism - which we call plugins) that allows to contribute compile time elements to the whole Theia code base. This isn't limited to new additions but allows to change virtually any part of the Framework. See [our documentation on the topic](https://theia-ide.org/docs/composing_applications/).

A good example to showcase the extension mechanism is the [Arduino IDE](https://github.com/arduino/arduino-ide), which exchanges large parts of the functionality of Theia, while not violating the EPL license in any way.

Note that I generally agree with you, that this is arguably the worst aspect of the EPL.

> Don't you think that returning to the Apache License or MIT would be better for this project? If not, why? What is the benefit of EPL for the project and its contributors?

Finally coming to the license question: It's a historic matter. Until 2021, the Eclipse Foundation enforced the use of the EPL in all projects. By now, this restriction has lifted. However, relicensing a project is no easy matter even if we wanted to. It would require agreement of most contributing parties, with a lot of them no longer active on the project. If Theia were a new Eclipse Project, it would likely stick with Apache or MIT. See also the accompanying [Theia-IDE](https://github.com/eclipse-theia/theia-ide) project which only recently [became completely MIT](https://github.com/eclipse-theia/theia-ide/issues/310). Newer projects contributed to the Eclipse Foundation are now mostly using MIT, see some of the [other](https://github.com/eclipse-langium/langium) [projects](https://github.com/eclipse-oct/open-collaboration-tools) I'm involved in.

> someone who is considering investing my time in either Theia or VS Code (including my possible contributions).

If I might add an argument: We're part of an open source foundation, so any contributions are legally ensured to always stay open and accessible to everyone. VS Code is completely owned by Microsoft and they can do whatever they want with the project (which is good for them, but not for the open source community).

**Reply by nimo23** [reply]
@msujew Thanks for clarifying. 

**To sum it up, and please correct me if the following statement is wrong:**

`Theia IDE` is `MIT` licensed, while `Theia framework` is `EPL` licensed. `Theia IDE` uses the framework, but is **NOT** a **derivative work**, as any "modifications" or "customizations" are only done via custom _extensions_ (or _subclasses_). So if I want to **customize/adopt** a React component or CSS style from the `EPL`-source, I can do that in **my own code** (e.g., by **extending a subclass** or **adding a CSS style**). With such customizations, a proprietary work (or the final product in object code) can have any type of license and does not have to distribute the source code to third parties or distribute the source code of the EPL code. Correct?

**Reply by msujew** [reply, marked as answer]
> So if I want to customize/adopt a React component or CSS style from the EPL-source, I can do that in my own code (e.g., by extending a subclass or adding a CSS style). 

Yep, you can even change the existing CSS by importing your own CSS files and adding `!important` to your rules.

> With such customizations, a proprietary work (or the final product in object code) can have any type of license and does not have to distribute the source code to third parties or distribute the source code of the EPL code. Correct?

Yes exactly. The license only requires you to redistribute your changes in case you touch the EPL licensed files directly. All of your work that just uses the Theia code/framework is free to use any license and stay closed source.

Generally, as long as you simply consume Theia via the published npm packages (like we do in the Theia-IDE project) you are on the safe side. There are very few cases in which it really makes sense to change the Theia framework source files, in which case you are best served by contributing those changes back to this repository as a PR.

**Reply by JonasHelming** [reply]
I would like to add one point about the Theia IDE: The tool is basically packaging Theia Platform packages together, adds a few really IDE specific things (such as the updater). So the repo contains mainly product configurations of one specific tool based on theia (the theia IDE). It can be used as a blueprint to build your own tool in a similar way. For this particular use case, we expect that forking can be necessary (of the Theia IDE repo), e.g. forking the build scripts.
This is the main reason why we have re licensed the Theia IDE repo under MIT, so this use case is supported.
